### PR TITLE
[MM-20482] Migrate 'components/post_view/channel_intro_message' module …

### PR DESCRIPTION
**Pull request by cinlloc**
…and associated tests to TypeScript

#### Summary
Migrate 'components/post_view/channel_intro_message' module and associated tests to TypeScript.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/15459

### CODE REVIEW

**Comment by agnivade** [comment]
Removing myself as I am not the right person to review this.

File: components/post_view/channel_intro_message/channel_intro_message.tsx [hunk 1]
```diff
@@ -283,7 +287,7 @@ export function createDefaultIntroMessage(channel, centeredIntro, enableUserCrea
                             {(title) => (
                                 <i
                                     className='fa fa-user-plus'
-                                    title={title}
+                                    title={title as string}
```
Comment (agarciamontoro) [inline components/post_view/channel_intro_message/channel_intro_message.tsx:290]: Why do we need this type casting here and in the change above? Can't we add the type to this lambda's parameter?

File: components/post_view/channel_intro_message/channel_intro_message.tsx [hunk 2]
```diff
@@ -438,7 +442,9 @@ function createStandardIntroMessage(channel, centeredIntro, locale, creatorName)
         );
     }
 
-    var purposeMessage = '';
+    let purposeMessage: JSX.Element = (
+        <span/>
+    );
```
Comment (agarciamontoro) [inline components/post_view/channel_intro_message/channel_intro_message.tsx:447]: Can't Typescript assume the type here?

File: components/post_view/channel_intro_message/channel_intro_message.tsx [hunk 3]
```diff
@@ -531,7 +537,7 @@ function createInviteChannelButton(channel) {
                     {(title) => (
                         <i
                             className='fa fa-user-plus'
-                            title={title}
+                            title={title as string}
```
Comment (agarciamontoro) [inline components/post_view/channel_intro_message/channel_intro_message.tsx:540]: Same question as before, about adding the type to the parameter instead of type casting (if possible).

File: components/post_view/channel_intro_message/index.ts [hunk 4]
```diff
@@ -9,29 +9,33 @@ import {getCurrentTeam} from 'mattermost-redux/selectors/entities/teams';
...
 
     return {
         currentUserId: getCurrentUserId(state),
         channel,
+        fullWidth: get(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.CHANNEL_DISPLAY_MODE, Preferences.CHANNEL_DISPLAY_MODE_DEFAULT) === Preferences.CHANNEL_DISPLAY_MODE_FULL_SCREEN,
         locale: getCurrentLocale(state),
         channelProfiles: getProfilesInCurrentChannel(state),
         enableUserCreation,
         isReadOnly,
-        fullWidth: get(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.CHANNEL_DISPLAY_MODE, Preferences.CHANNEL_DISPLAY_MODE_DEFAULT) === Preferences.CHANNEL_DISPLAY_MODE_FULL_SCREEN,
```
Comment (agarciamontoro) [inline components/post_view/channel_intro_message/index.ts:34]: Just out of curiosity: is this change in the order needed? I don't have anything against it, but I guess that it's not needed because it's a named field.

File: package-lock.json [hunk 5]
```diff
@@ -12582,15 +12582,11 @@
       "resolved": "https://registry.npmjs.org/full-icu/-/full-icu-1.3.1.tgz",
       "integrity": "sha512-VMtK//85QJomhk3cXOCksNwOYaw1KWnYTS37GYGgyf7A3ajdBoPGhaJuJWAH2S2kq8GZeXkdKn+3Mfmgy11cVw==",
       "dev": true,
-      "requires": {
-        "icu4c-data": "0.64.2"
-      },
```
Comment (agarciamontoro) [inline package-lock.json:12587]: Is this removal related to the PR?

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
This is great, @cinlloc! Thank you for your contribution :tada: I left a couple of comments below, mostly questions, more than changes requested.

File: package-lock.json (hunk 5 above)
Comment (cinlloc) [reply package-lock.json:12587]: Not really, I remove it.

File: components/post_view/channel_intro_message/index.ts (hunk 4 above)
Comment (cinlloc) [reply components/post_view/channel_intro_message/index.ts:34]: Not needed indeed. It moved it for readability purpose during work. I can revert this if you feel so.

File: components/post_view/channel_intro_message/channel_intro_message.tsx (hunk 1 above)
Comment (cinlloc) [reply components/post_view/channel_intro_message/channel_intro_message.tsx:290]: You're right, I added the type to the lambda parameter and removed the cast.

File: components/post_view/channel_intro_message/channel_intro_message.tsx (hunk 2 above)
Comment (cinlloc) [reply components/post_view/channel_intro_message/channel_intro_message.tsx]: Yes, it works without, I fix it.

File: components/post_view/channel_intro_message/index.ts (hunk 4 above)
Comment (agarciamontoro) [reply components/post_view/channel_intro_message/index.ts:34]: No need to do so, it's ok!

**Review by agarciamontoro — APPROVED** [review]
Great job, @cinlloc, thank you for your contribution :tada:

**Review by iomodo — APPROVED** [review]
LGTM! Thanks for the contribution @cinlloc

**Review by josephbaylon — CHANGES_REQUESTED** [review]
Hi @cinlloc Thanks for your contribution !! Please add unit tests for `channel_intro_message`. Here's the guide to writing unit tests: https://developers.mattermost.com/contribute/webapp/unit-testing/

Also, as a bonus, it would also be good to add a happy path E2E for the same. Here's the guide to writing e2e tests:
https://developers.mattermost.com/contribute/webapp/end-to-end-tests/

**Comment by josephbaylon** [comment]
@cinlloc let us know if we can help you somehow.

**Review by josephbaylon — APPROVED** [review]
LGTM. I just checked again and verified channel intro message is being tested from `post_list_row.test.tsx` and also ran some e2e tests that cover channel intro. So this should be good now @cinlloc sorry for the confusion.

**Comment by josephbaylon** [comment]
@cinlloc Please resolve conflicts

**Comment by cinlloc** [comment]
Conflicts solved.

**Comment by hanzei** [comment]
Thanks for the great PR @cinlloc :man_dancing:
